### PR TITLE
Add Document and Index classes

### PR DESCRIPTION
**Pull request by vidok**
## Closes https://github.com/elastic/enterprise-search-team/issues/3252

- [ ] BYOC refactoring (Adding Index and Document classes)
- [ ] Job processing (requesting jobs index to perform a job)

#### Pre-Review Checklist
- [ ] Covered the changes with automated tests
- [ ] Tested the changes locally

#### Changes Requiring Extra Attention

- [ ] Security-related changes (encryption, TLS, SSRF, etc)
- [ ] New external service dependencies added.

### CODE REVIEW

File: connectors/byoc.py [hunk 1]
```diff
@@ -64,14 +66,15 @@ class JobStatus(Enum):
 )
 
 
-class BYOIndex(ESClient):
+class BYOIndex(Index):
     def __init__(self, elastic_config):
         super().__init__(elastic_config)
```
Comment (tarekziade) [inline connectors/byoc.py:71]: Should be 

```
super().__init__(CONNECTORS_INDEX, elastic_config)
```

File: connectors/utils/document.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class Document:
+  def __init__(self, elastic_index, id, doc_source):
+    self.elastic_index = elastic_index
+    self.id = id
+    self.doc_source = doc_source
+
+  async def reload(self):
+    resp = await self.elastic_client.get(
```
Comment (tarekziade) [inline connectors/utils/document.py:19]: you can `return await ...` directly

File: connectors/utils/index.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    raise NotImplementedError
+
+  async def get_docs(self):
+    await self.client.indices.refresh(index=self.index_name)
+
+    try:
+        resp = await self.client.search(
+            index=self.index_name,
+            query={"match_all": {}},
+            size=20,
```
Comment (tarekziade) [inline connectors/utils/index.py:31]: let's make `20` a value we grab from `elastic_config` with a default size `DEFAULT_SEARCH_SIZE=20`

File: connectors/utils/index.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            expand_wildcards="hidden",
+        )
+    except ApiError as e:
+        logger.critical(f"The server returned {e.status_code}")
+        logger.critical(e.body, exc_info=True)
+        return
+
+    logger.debug(f"Found {len(resp['hits']['hits'])} documents")
+    for doc in resp["hits"]["hits"]:
+      yield self._create_object(doc)
```
Comment (tarekziade) [inline connectors/utils/index.py:41]: run `make lint` (we want 4 space indents)

**Review by tarekziade — COMMENTED** [review]
great structure! I added a few comments for your py-ramp up :)

File: connectors/byoc.py [hunk 5]
```diff
@@ -20,6 +20,8 @@
 from connectors.source import DataSourceConfiguration
 from elasticsearch.exceptions import ApiError
 from connectors.index import defaults_for, DEFAULT_LANGUAGE
+from connectors.utils.index import Index
```
[the remark this answers is not in the document]
Comment (vidok) [reply connectors/byoc.py]: Good point. However, I would keep the `Es` prefix for the classes`s names (`EsIndex`,`EsDocument`)

File: connectors/utils.py [hunk 6]
```diff
@@ -300,3 +300,52 @@ async def _wait_for_room(self, item):
     async def put(self, item):
         item_size = await self._wait_for_room(item)
         return await super().put((item_size, item))
+
+
+class EsDocument:
```
Comment (tarekziade) [inline connectors/utils.py:305]: nit: pep8 we'd probably use `ESDocument`

**Comment by timgrein** [comment]
cc @tarekziade @wangch079 can we close this PRs? Just recognizing that we currently always sit at ~ 15 PRs and many of them seem to be stale?

**Comment by wangch079** [comment]
>cc @tarekziade @wangch079 can we close this PRs? Just recognizing that we currently always sit at ~ 15 PRs and many of them seem to be stale?

I do think we could close this. Both `ESDocument` and `ESIndex` classes are covered in other PRs. cc @vidok

**Pull request closed without merging**